core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -572,14 +579,19 @@ private[log] class Cleaner(val id: Int,
         val abortedTransactions = log.collectAbortedTransactions(startOffset, upperBoundOffset)
         transactionMetadata.addAbortedTransactions(abortedTransactions)
 
-        val retainDeletesAndTxnMarkers = currentSegment.lastModified > deleteHorizonMs
+        val retainDeletesAndTxnMarkers = currentSegment.lastModified > trackedHorizon
         info(s"Cleaning $currentSegment in log ${log.name} into ${cleaned.baseOffset} " +
-          s"with deletion horizon $deleteHorizonMs, " +
+          s"with deletion horizon $trackedHorizon, " +
           s"${if(retainDeletesAndTxnMarkers) "retaining" else "discarding"} deletes.")
 
         try {
-          cleanInto(log.topicPartition, currentSegment.log, cleaned, map, retainDeletesAndTxnMarkers, log.config.maxMessageSize,
-            transactionMetadata, lastOffsetOfActiveProducers, stats)
+          val latestDeleteHorizon: Long = cleanInto(log.topicPartition, currentSegment.log, cleaned, map, retainDeletesAndTxnMarkers, log.config.deleteRetentionMs,
+                    log.config.maxMessageSize, transactionMetadata, lastOffsetOfActiveProducers, stats, currentTime = currentTime)
+          if (log.latestDeleteHorizon < latestDeleteHorizon) {
+            log.latestDeleteHorizon = latestDeleteHorizon
+          } else if (log.latestDeleteHorizon < currentTime) {
+            log.latestDeleteHorizon = RecordBatch.NO_TIMESTAMP
```

Comment:
I am not sure that I follow the logic here. To me, the easiest way is to reset log.latestDeleteHorizon at the beginning of each round of cleaning. Then, we update it with the latestDeleteHorizon remaining in each cleaned segment.